File: python/difference-of-squares/difference_of_squares.py
"""
Find the difference between the square of the sum and the
sum of the squares of the first N natural numbers.
"""


def square_of_sum(number):
    """
    :param number: int number to calculate
    :return: int calculated value
    """

    total = sum(range(1, number + 1))
    return total * total


def sum_of_squares(number):
    """
    :param number: int number to calculate
    :return: int calculated value
    """
    return sum(x * x for x in range(1, number + 1))


def difference_of_squares(number):
    """
    :param number: int number to calculate
    :return: int calculated value
    """

    return square_of_sum(number) - sum_of_squares(number)

File: ocaml/difference-of-squares/difference_of_squares.ml
(*
Find the difference between the square of the sum and the sum of the squares of the first N natural numbers.

The square of the sum of the first ten natural numbers is (1 + 2 + ... + 10)² = 55² = 3025.

The sum of the squares of the first ten natural numbers is 1² + 2² + ... + 10² = 385.

Hence the difference between the square of the sum of the first ten natural
numbers and the sum of the squares of the first ten natural numbers is 3025 -
385 = 2640.
*)

open Base

let sum l = List.fold ~init:0 ~f:( + ) l
let rec range a b = if a > b then [] else a :: range (a + 1) b

let square_of_sum n =
  let total = range 1 n |> sum in
  total * total

let sum_of_squares n = range 1 n |> List.map ~f:(fun x -> x * x) |> sum
let difference_of_squares n = square_of_sum n - sum_of_squares n
